### PR TITLE
Added errorSelector field for custom errors

### LINKED ISSUE #11387: Add selector for custom errors in compiler output
## Abstract

Add the selector of a custom error to the `sources.contract.ast` entry of the compiler output, similar to how functions have a `functionSelector` property.

## Motivation

To detect which custom error was thrown, you need to inspect the return data of a revert and check the first four bytes to see which signature they match. If a tool wants to use the compiler output to detect this automatically (like Hardhat does), right now you'll need to manually build the signature and hash it. For simple types this is not that hard, but if the signature includes nested data with arrays and or structs, things get way harder. 

## Specification

I'm not sure about this part. I guess `errorSelector` would work, analog to the `functionSelector` one. I also wonder if it would make sense to have an `errorIdentifiers` entry, similar to the current `methodIdentifiers`.

Also, and maybe unrelated to this issue: the same thing happens with events. Since as far as I know custom errors are somewhat modeled after events, maybe it would make sense to include them as part of this change. We don't need that right now, but I think they should be in the output too. (If that would make more sense as a separate issue, please let me know and I'll submit it).

## Backwards Compatibility

I think adding a new property in the compiler output is backwards compatible.

### PR DESCRIPTION
**Pull request by nishant-sachdeva**
Closes #11387

### CODE REVIEW

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 1]
```diff
@@ -404,7 +404,8 @@ bool ASTJsonConverter::visit(FunctionDefinition const& _node)
 		make_pair("modifiers", toJson(_node.modifiers())),
 		make_pair("body", _node.isImplemented() ? toJson(_node.body()) : Json::nullValue),
 		make_pair("implemented", _node.isImplemented()),
-		make_pair("scope", idOrNull(_node.scope()))
+		make_pair("scope", idOrNull(_node.scope())),
+		make_pair("errorSelector", _node.functionType(true)->externalIdentifierHex())
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonConverter.cpp]: This should go to `ErrorDefinition`

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 2]
```diff
@@ -404,7 +404,7 @@ bool ASTJsonConverter::visit(FunctionDefinition const& _node)
 		make_pair("modifiers", toJson(_node.modifiers())),
 		make_pair("body", _node.isImplemented() ? toJson(_node.body()) : Json::nullValue),
 		make_pair("implemented", _node.isImplemented()),
-		make_pair("scope", idOrNull(_node.scope()))
+		make_pair("scope", idOrNull(_node.scope())),
```
Comment (ekpyron) [inline libsolidity/ast/ASTJsonConverter.cpp]:
```suggestion
		make_pair("scope", idOrNull(_node.scope()))
```

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 3]
```diff
@@ -498,7 +498,8 @@ bool ASTJsonConverter::visit(EventDefinition const& _node)
 		make_pair("nameLocation", sourceLocationToString(_node.nameLocation())),
 		make_pair("documentation", _node.documentation() ? toJson(*_node.documentation()) : Json::nullValue),
 		make_pair("parameters", toJson(_node.parameterList())),
-		make_pair("anonymous", _node.isAnonymous())
+		make_pair("anonymous", _node.isAnonymous()),
+		make_pair("eventSelector", _node.functionType(true)->externalIdentifierHex())
```
Comment (ekpyron) [inline libsolidity/ast/ASTJsonConverter.cpp]: Is this really what we want to output? Compare with https://github.com/ethereum/solidity/pull/12548.

File: test/libsolidity/ASTJSON/used_errors.json [hunk 4]
```diff
@@ -1,231 +0,0 @@
-{
```
Comment (ekpyron) [inline test/libsolidity/ASTJSON/used_errors.json]: This file being removed happens due to the test file containing analysis errors - I recently reported this as https://github.com/ethereum/solidity/issues/12621
But in this case here: is it intentional that there are analysis errors in this test? Probably not (I haven't looked at the input myself)? Then we should fix them.

**Review by ekpyron — COMMENTED** [review]
Some comments - also needs a changelog entry.

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 5]
```diff
@@ -493,24 +493,32 @@ bool ASTJsonConverter::visit(ModifierInvocation const& _node)
...
-	setJsonNode(_node, "EventDefinition", {
+	std::vector<pair<string, Json::Value>> _attributes = {
 		make_pair("name", _node.name()),
 		make_pair("nameLocation", sourceLocationToString(_node.nameLocation())),
 		make_pair("documentation", _node.documentation() ? toJson(*_node.documentation()) : Json::nullValue),
 		make_pair("parameters", toJson(_node.parameterList())),
 		make_pair("anonymous", _node.isAnonymous())
-	});
+	};
+	if (m_stackState > CompilerStack::State::ParsedAndImported)
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonConverter.cpp]: Can you use `>=` for comparison? We need type checking here, don't we?

Comment (chriseth) [reply libsolidity/ast/ASTJsonConverter.cpp]: I'm not sure if `AnalysisPerformed` is acutally enough as well, because it is set even if there were errors.

File: test/libsolidity/ASTJSON/documentation.json [hunk 6]
```diff
@@ -162,6 +162,7 @@
             "src": "69:26:3",
             "text": "Some comment on Evt."
           },
+          "eventSelector": "a6900791",
```
Comment (chriseth) [inline test/libsolidity/ASTJSON/documentation.json]: This is too short (just to keep an open note)

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 7]
```diff
@@ -493,24 +493,32 @@ bool ASTJsonConverter::visit(ModifierInvocation const& _node)
...
+	std::vector<pair<string, Json::Value>> _attributes = {
 		make_pair("name", _node.name()),
 		make_pair("nameLocation", sourceLocationToString(_node.nameLocation())),
 		make_pair("documentation", _node.documentation() ? toJson(*_node.documentation()) : Json::nullValue),
 		make_pair("parameters", toJson(_node.parameterList())),
 		make_pair("anonymous", _node.isAnonymous())
-	});
+	};
+	if (m_stackState >= CompilerStack::State::AnalysisPerformed)
+			_attributes.emplace_back(make_pair("eventSelector", _node.functionType(true)->externalIdentifierHex()));
```
Comment (ekpyron) [inline libsolidity/ast/ASTJsonConverter.cpp]:
```suggestion
			_attributes.emplace_back(make_pair("eventSelector", toHex(u256(h256::Arith(keccak256(_node.functionType(true)->externalSignature())))));
```
Something like this.

File: test/libsolidity/ASTJSON/used_errors.json (hunk 4 above)
Comment (nishant-sachdeva) [reply test/libsolidity/ASTJSON/used_errors.json]: I restored the contents of the file after recent code changes ( as an experiment ) and the tests  ran perfectly. Have added in the latest push to the PR

**Review by Marenz — CHANGES_REQUESTED** [review]
I don't see any test case result that includes `errorSelector`. We should have at least one test for that

**Comment by chriseth** [comment]
Please add a changelog entry.

File: Changelog.md [hunk 8]
```diff
@@ -2,12 +2,14 @@
 
 Language Features:
 
-
 Compiler Features:
-
+* General: Added selector field in the Error and Event Description.
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
* JSON-AST: Added selector field for errors and events.
```

File: Changelog.md [hunk 9]
```diff
@@ -2,12 +2,14 @@
...
 
 Bugfixes:
 
+Solc-Js:
```
Comment (chriseth) [inline Changelog.md]: Please remove.

**Comment by chriseth** [comment]
Can you please add a test where you have an event that uses internal types like an internal function? For example `event E(function() internal);` - this should generate the AST json, but without an error selector.

File: Changelog.md [hunk 10]
```diff
@@ -2,14 +2,12 @@
 
 Language Features:
 
-
```
Comment (chriseth) [inline Changelog.md:5]: Please keep the double empty lines.

File: test/libsolidity/ASTJSON/used_errors.sol [hunk 11]
```diff
@@ -2,7 +2,7 @@ error X();
 function f() { revert X(); }
 contract C {
     error T();
-    function h() public pure { f(); }
+    function h() public { f(); }
```
Comment (chriseth) [inline test/libsolidity/ASTJSON/used_errors.sol:5]: Why did you remove the pure?

**Review by chriseth — CHANGES_REQUESTED** [review]
Please don't forget to add the suggested test case.

File: test/libsolidity/ASTJSON/used_errors.sol (hunk 11 above)
Comment (nishant-sachdeva) [reply test/libsolidity/ASTJSON/used_errors.sol:5]: The code in this .sol file was giving errors on compilation.  Removing pure fixed them

**Review by chriseth — APPROVED** [review]
